haystack/nodes/answer_generator/openai.py

```diff
@@ -204,59 +259,73 @@ def predict(
         result = {"query": query, "answers": answers}
         return result
 
+    @staticmethod
+    def _create_context(documents: List[Document], join_str: str = " ") -> str:
+        """Join the documents together to create a single context to be used in the PromptTemplate."""
+        doc_contents = [doc.content for doc in documents]
+        context = join_str.join(reversed(doc_contents))
+        return context
+
     def _build_prompt(self, query: str, documents: List[Document]) -> Tuple[str, List[Document]]:
         """
         Builds the prompt for the GPT-3 model so that it can generate an Answer.
         """
-        example_context = f"===\nContext: {self.examples_context}\n===\n"
-        example_prompts = "\n---\n".join([f"Q: {question}\nA: {answer}" for question, answer in self.examples])
-        instruction = "Please answer the question according to the above context.\n" + example_context + example_prompts
-        instruction = f"{instruction.strip()}\n\n"
-
+        example_prompts = "\n---\n".join([f"Q: {query}\nA: {answer}" for query, answer in self.examples])
         qa_prompt = f"Q: {query}\nA:"
 
-        n_instruction_tokens = self._count_tokens(instruction + qa_prompt + "===\nContext: \n===\n")
-
-        logger.debug("Number of tokens in instruction: %s", n_instruction_tokens)
-
-        n_docs_tokens = [self._count_tokens(doc.content) for doc in documents]
-        logger.debug("Number of tokens in documents: %s", n_docs_tokens)
+        kwargs = {"context": self._create_context(documents, join_str=self.context_join_str), "query": qa_prompt}
+        if (
+            "examples_context" in self.instruction_prompt.prompt_params
+            and "examples" in self.instruction_prompt.prompt_params
+        ):
+            kwargs["examples_context"] = self.examples_context
+            kwargs["examples"] = example_prompts
+        full_prompt = next(self.instruction_prompt.fill(**kwargs))
+        n_full_prompt_tokens = self._count_tokens(full_prompt)
 
         # for length restrictions of prompt see: https://beta.openai.com/docs/api-reference/completions/create#completions/create-max_tokens
-        leftover_token_len = self.MAX_TOKENS_LIMIT - n_instruction_tokens - self.max_tokens
+        leftover_token_len = self.MAX_TOKENS_LIMIT - n_full_prompt_tokens - self.max_tokens
 
-        # Add as many Documents as context as fit into the model
-        input_docs = []
-        input_docs_content = []
+        input_docs = documents
         skipped_docs = 0
-        for doc, doc_token_len in zip(documents, n_docs_tokens):
-            if doc_token_len <= leftover_token_len:
-                input_docs.append(doc)
-                input_docs_content.append(doc.content)
-                leftover_token_len -= doc_token_len
-            else:
+        if leftover_token_len < 0:
```

Comment:
Please add some comments here to explain your key logic for future debuggers. If I get it right you reversed the initial logic and now rather trim down the prompt until it fits the model requirements.

Reply:
Yes, that is correct. I found this to be the easiest way now because it is more difficult to count how many tokens the entire prompt will take without first constructing it. Tokenizing the just the template (with the `$` variables) would result in an incorrect token count.